// @plugnet/types/codec/create/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TypeDefInfo = void 0;
// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
// Type which says: if `K` is in the InterfaceRegistry, then return InterfaceRegistry[K], else fallback to T
let TypeDefInfo;
exports.TypeDefInfo = TypeDefInfo;

(function (TypeDefInfo) {
  TypeDefInfo[TypeDefInfo["Compact"] = 0] = "Compact";
  TypeDefInfo[TypeDefInfo["DoubleMap"] = 1] = "DoubleMap";
  TypeDefInfo[TypeDefInfo["Enum"] = 2] = "Enum";
  TypeDefInfo[TypeDefInfo["Linkage"] = 3] = "Linkage";
  TypeDefInfo[TypeDefInfo["Option"] = 4] = "Option";
  TypeDefInfo[TypeDefInfo["Plain"] = 5] = "Plain";
  TypeDefInfo[TypeDefInfo["Set"] = 6] = "Set";
  TypeDefInfo[TypeDefInfo["Struct"] = 7] = "Struct";
  TypeDefInfo[TypeDefInfo["Tuple"] = 8] = "Tuple";
  TypeDefInfo[TypeDefInfo["Vec"] = 9] = "Vec";
  TypeDefInfo[TypeDefInfo["VecFixed"] = 10] = "VecFixed";
  TypeDefInfo[TypeDefInfo["Null"] = 11] = "Null";
})(TypeDefInfo || (exports.TypeDefInfo = TypeDefInfo = {}));
